### PR TITLE
Enable zero-shot learning with CLIP

### PR DESCRIPTION
**Pull request by zhiqiangdon**
Zero-shot API examples:
```
from PIL import Image
import requests

from autogluon.multimodal import MultiModalPredictor

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
cat_image_name = "cat.jpg"
image.save(cat_image_name)

url = "https://farm1.staticflickr.com/29/57154382_07b25134f7_z.jpg"
image = Image.open(requests.get(url, stream=True).raw)
dog_image_name = "dog.jpg"
image.save(dog_image_name)

cat_text = "a photo of a cat"
dog_text = "a photo of a dog"
bird_text = "a photo of a bird"

predictor = MultiModalPredictor(hyperparameters={"model.names": ["clip"]}, problem_type="zero_shot")

# compute the cosine similarity of one image-text pair.
pred = predictor.predict({"image": [cat_image_name], "text": [cat_text]})
print(f"compute the cosine similarity of one image-text pair: {pred}")

# compute the cosine similarities of more image and text pairs.
pred = predictor.predict({"image": [cat_image_name, dog_image_name], "text": [cat_text, dog_text]})
print(f"compute the cosine similarities of more image and text pairs: {pred}")

# match images in a given text pool and output the matched text index (starting from 0) for each image.
pred = predictor.predict({"image0": [cat_image_name, dog_image_name]}, {"names": [dog_text, cat_text, bird_text]})
print(f"match images in a given text pool and output the matched text index (starting from 0) for each image: {pred}")

# match texts in a given image pool and output the matched image index (starting from 0) for each text.
pred = predictor.predict({"query": [dog_text, cat_text, bird_text]}, {"candidates": [cat_image_name, dog_image_name]})
print(f"match texts in a given image pool and output the matched image index (starting from 0) for each text: {pred}")

# predict the probabilities of one image matching several texts.
prob = predictor.predict_proba({"image": [cat_image_name]}, {"text": [cat_text, dog_text, bird_text]})
print(f"predict the probabilities of one image matching several texts: {prob}")

# given two or more images, we can get the probabilities of matching each image with a pool of texts.
prob = predictor.predict_proba({"image": [dog_image_name, cat_image_name]}, {"text": [bird_text, cat_text, dog_text]})
print(f"given two or more images, we can get the probabilities of matching a pool of texts: {prob}")

# predict the probabilities of one text matching several images.
prob = predictor.predict_proba({"x": [cat_text]}, {"y": [dog_image_name, cat_image_name]})
print(f"predict the probabilities of one text matching several images: {prob}")

# given two or more texts, we can get the probabilities of matching each text with a pool of images.
prob = predictor.predict_proba({"a": [bird_text, cat_text, dog_text]}, {"b": [dog_image_name, cat_image_name]})
print(f"given two or more texts, we can get the probabilities of matching a pool of images: {prob}")

# extract image embeddings.
embedding = predictor.extract_embedding({"123": [dog_image_name, cat_image_name]})
for k, v in embedding.items():
    print(k, v.shape)

# extract text embeddings.
embedding = predictor.extract_embedding({"xyz": [bird_text, dog_text, cat_text]})
for k, v in embedding.items():
    print(k, v.shape)

# extract embeddings for both images and texts.
embedding = predictor.extract_embedding({"image": [cat_image_name, dog_image_name], "text": [bird_text, dog_text]})
for k, v in embedding.items():
    print(k, v.shape)



# invalid API usage 1: passing one dictionary, but different keys have inconsistent list lengths.
pred = predictor.predict({"image": [cat_image_name, dog_image_name], "text": [cat_text]})

embedding = predictor.extract_embedding({"image": [cat_image_name], "text": [cat_text, bird_text]})

# invalid API usage 2: predicting probability with only one dictionary as input.
prob = predictor.predict_proba({"image": [cat_image_name], "text": [cat_text]})
```

To support zero-shot learning:

1. Add the `hyperparameters` argument in `MultiModalPredictor`'s init function.
2. A new problem type `zero_shot`.
3. Refactor `infer_column_problem_types` into three functions: `infer_column_types`, `infer_label_column_type_by_problem_type`, and `infer_problem_type_output_shape`. The `infer_column_types` supports using only data.
4. Make data processors no longer depend on df_preprocessor. Now we can create data processors simply from model configs.
5. Support initializing df_preprocessor without data.
6. Make CLIP return logits (cosine similarity) in the zero-shot setting.
7. Support incomplete input (either text or image) for CLIP. This is necessary to extract single modality embeddings.
8. Add an API to extract desired information from prediction outputs.
9. Add `_on_predict_start()` for prediction.
10. Add zero-shot initialization.
11. Zero-shot learning unit tests.

Others:

1. Due to the upgrade of `nlpaug` (https://github.com/makcedward/nlpaug), specifying `model.hf_text.text_train_augment_types` with a list of augmentation operators causes exceptions. To workaround this, we modify the hyperparameters in `test_textagumentor_deepcopy`.
2. Correct the `test_textagumentor_deepcopy` to deepcopy `data_processors` instead of `df_preprocessor`.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/data/infer_types.py [hunk 1]
```diff
@@ -326,7 +355,7 @@ def infer_problem_type_output_shape(
     """
     if provided_problem_type is not None:
         if provided_problem_type == MULTICLASS or provided_problem_type == BINARY:
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/infer_types.py:357]: In the following, we also tested for `provided_problem_type == REGRESSION`. Actually, it will never reach this condition because we tested via `if provided_problem_type == MULTICLASS or provided_problem_type == BINARY:`. Thus, we should change the if condition to 

```python
if provided_problem_type in [MULTICLASS, BINARY, REGRESSION]:
```

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py [hunk 2]
```diff
@@ -265,7 +293,6 @@ def transform_text(
         -------
         All the text data stored in a dictionary.
         """
-        assert self._fit_called, "You will need to first call preprocessor.fit before calling preprocessor.transform."
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:268]: Should we remove the assertion here? We can still call `.fit()` even if it's the zero-shot scenario. We may just ensure that `y=None` when calling .fit(). For example, `.fit(X, Y=None)`.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 3]
```diff
@@ -1537,6 +1606,11 @@ def extract_embedding(
         data
             The data to extract embeddings for. Should contain same column names as training dataset and
             follow same format (except for the `label` column).
+        return_masks
+            Whether to return the masks indicating whether the corresponding features are valid.
+            If one sample's some column have empty input, then the sample has mask 0 for that column.
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/predictor.py:1611]: Consider to change to 

```python
Return the masks that indicate the sample contains missing values. If a sample contains missing values, the features can be invalid and we use normalized constant vectors as the feature.
```

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 2 above)
Comment (bryanyzhu) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:268]: There is another situation (maybe out of scope), that is doing self-supervised learning. Take contrastive learning as an example, we only have `X` as input, both positive and negative samples are from the same `X`. So there is no `Y`

**Review by bryanyzhu — COMMENTED** [review]
Thanks for the PR! This is super important.

File: multimodal/src/autogluon/multimodal/data/infer_types.py (hunk 1 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/infer_types.py:357]: `if provided_problem_type == REGRESSION` isn't under the condition `if provided_problem_type == MULTICLASS or provided_problem_type == BINARY`. It's under the condition `if provided_problem_type is not None`.

File: multimodal/src/autogluon/multimodal/models/clip.py [hunk 4]
```diff
@@ -115,71 +115,81 @@ def forward(
...
+            ret[COLUMN_FEATURES][MASKS].update(text_column_feature_masks)
+
+        if has_image and has_text:
+            if self.num_classes:
+                features = image_features + text_features
+                logits = self.head(features)
+                ret[FEATURES] = features
+            else:
+                # cosine similarity as logits
+                logits = torch.sum(image_features * text_features, dim=-1)
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/clip.py:190]: The features are normalized in above.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1611]: The mask is feature column-wise, not sample-wise. A sample may have two feature columns `image1` and `image2`. We extract features for `image1` and `image2` separately. Maybe the image of `image1` is missing, but `image2` has a valid image. Then `image1` of this sample has mask 0 while `image2` of this sample has mask 1.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:1611]: Okay, the description is not very easy to understand. Also, do we really need it?

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1611]: Users may need this information in some cases. Returning masks is optional (default False).

File: multimodal/src/autogluon/multimodal/data/infer_types.py [hunk 5]
```diff
@@ -218,88 +215,120 @@ def infer_column_problem_types(
...
+        ):  # Infer categorical column
             column_types[col_name] = CATEGORICAL
-        elif is_numerical_column(train_df[col_name], valid_df[col_name]):
+        elif is_numerical_column(data[col_name], valid_data[col_name]):  # Infer numerical column
             column_types[col_name] = NUMERICAL
-        elif is_imagepath_column(train_df[col_name], col_name):
+        elif is_imagepath_column(data[col_name], col_name):  # Infer image-path column
             column_types[col_name] = IMAGE_PATH
-        elif check_if_nlp_feature(train_df[col_name]):
+        elif check_if_nlp_feature(data[col_name]):  # Infer text column
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/infer_types.py:254]: Good catch. This name was from `autogluon.core`. Will rename it to make it consistent with others.

File: multimodal/src/autogluon/multimodal/models/clip.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/models/clip.py:190]: Yeah, I misread the code.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:1611]: We may need to improve the documentation. Currently it's difficult to understand this flag.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 2 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:268]: Calling `fit()` has no effect on image and text data. It is mainly for the tabular data. That's why I remove the assertions in the `transform_image()` and `transform_text()`.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:268]: Yeah, later on, we may consider to do something for Image/Text. Removing the assertion is fine for now (since our focus is to add more functionality) but we may consider to refactor the code later.

File: multimodal/tests/unittests/test_zero_shot.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+def test_clip_zero_shot():
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    image = Image.open(requests.get(url, stream=True).raw)
+    image_name = "clip_image_0.jpg"
+    image.save(image_name)
+
+    predictor = MultiModalPredictor(hyperparameters={"model.names": ["clip"]}, problem_type="zero_shot")
+
+    pred = predictor.predict({"image": [image_name], "text": ["a photo of a dog"]})
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/tests/unittests/test_zero_shot.py]: Good suggestions. Added more comments.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py]: Refined.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 2 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:268]: Refactored `fit()` into `fit_x()` and `fix_y()` and added back the assertions.

File: docs/tutorials/multimodal/beginner_text.md [hunk 7]
```diff
@@ -87,17 +87,17 @@ And you can easily obtain predictions from these models by calling `predictor.pr
 sentence1 = "it's a charming and often affecting journey."
 sentence2 = "It's slow, very, very, very slow."
 predictions = predictor.predict({'sentence': [sentence1, sentence2]})
-print('"Sentence":', sentence1, '"Predicted Sentiment":', predictions.iloc[0])
-print('"Sentence":', sentence2, '"Predicted Sentiment":', predictions.iloc[1])
+print('"Sentence":', sentence1, '"Predicted Sentiment":', predictions[0])
```
Comment (sxjscience) [inline docs/tutorials/multimodal/beginner_text.md:90]: iloc[0] means to take the 0-th element. Is there a specific reason for this change?

**Comment by szha** [comment]
Job PR-1922-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1922/11/index.html

File: docs/tutorials/multimodal/beginner_text.md (hunk 7 above)
Comment (zhiqiangdon) [reply docs/tutorials/multimodal/beginner_text.md:90]: To accommodate the zero-shot tutorial needs, `predict()` returns a numpy array if the input data is in the dictionary or list format. It returns a dataframe if the input data is a dataframe or users specify `as_pandas=True`.

Comment (sxjscience) [reply docs/tutorials/multimodal/beginner_text.md:90]: I see. Will it make sense to just change the default of "as_pandas"? Or we can leave it to a later PR.

Comment (sxjscience) [reply docs/tutorials/multimodal/beginner_text.md:90]: After a 2nd thought, I think the current design of having as_pandas=None by default also makes sense.

**Review by sxjscience — APPROVED** [review]
I think this change looks good to me. We can merge once it passed the CI.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py [hunk 8]
```diff
@@ -239,6 +262,21 @@ def fit(
                 raise NotImplementedError(
                     f"Type of the column is not supported currently. Received {col_name}={col_type}."
                 )
+
+    def fit_y(self, y: pd.Series):
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:266]: One caveat is that `fit_x` and `fit_y` may better be combined still. We can allow Y=None in `fit()`. This aligns with the design of sklearn.

However, I'm approving the PR now because I feel that we can refactor the code in a later PR.

File: docs/tutorials/multimodal/beginner_text.md (hunk 7 above)
Comment (zhiqiangdon) [reply docs/tutorials/multimodal/beginner_text.md:90]: The `predictions ` in the above case is a ndarray since the input is a dictionary. It is necessary to remove `iloc` here.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 8 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py]: Added back  `fit()`, which calls `_fit_x()` and `_fit_y()` internally.

**Review by bryanyzhu — APPROVED** [review]

**Comment by szha** [comment]
Job PR-1922-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1922/12/index.html

**Comment by szha** [comment]
Job PR-1922-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1922/13/index.html

**Comment by szha** [comment]
Job PR-1922-14 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1922/14/index.html
